### ISSUE TITLE
bug: Gitpod needs better way to handle access controls

### ISSUE THREAD

**Issue by Kreyren** [gitpod-io/gitpod#1055]
### TL;DR
Github rejects gitpod's publickey on organizations which didn't approve gitpod in access controls

Gitpod can be still opened on those organization repositories which causes issues with `git` -> Proposing to prompt up the end user on affected repositories for proper configuration to avoid confusion

### Describe the bug

I keep getting this issue on repositories which didn't approved gitpod application.

```sh
gitpod /workspace/kreypi $ git push
remote: Permission to kreypi/kreypi.git denied to Kreyren.
fatal: unable to access 'https://github.com/kreypi/kreypi.git/': The requested URL returned error: 403
gitpod /workspace/kreypi $ git^C
gitpod /workspace/kreypi $ touch test
gitpod /workspace/kreypi $ git add test
gitpod /workspace/kreypi $ git commit -m test
[master 85cc6fa] test
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 test
gitpod /workspace/kreypi $ git push
remote: Permission to kreypi/kreypi.git denied to Kreyren.
fatal: unable to access 'https://github.com/kreypi/kreypi.git/': The requested URL returned error: 403
gitpod /workspace/kreypi $ 
```

Proposing some check to be executed on each gitpod run for this issue and in case gitpod doesn't have the required access to push changes in the repository to warn the end-user and prompt with assigning said permission.

### To Reproduce

 Create a new organization with any repository on a GitHub account with access controls revoking access to the gitpod then open gitpod on that repository and try to push changes expecting failure due to the publickey rejection. 

### Expected behavior

Failure to push changes is not expected should prompt up the end-user if gitpod is executed on affected repositories to approve gitpod in access controls to avoid confusion and misconfiguration. 

### Additional Information

Assigning gitpod to have said permission through https://github.com/settings/connections/applications/484069277e293e6d2a2a fixes the issue.

<img width="674" alt="image" src="https://user-images.githubusercontent.com/11302521/71932556-55aab280-31a0-11ea-83ef-75b3e9fe77e3.png">

```console
gitpod /workspace/kreypi $ git push
remote: Permission to kreypi/kreypi.git denied to Kreyren.
fatal: unable to access 'https://github.com/kreypi/kreypi.git/': The requested URL returned error: 403
# After granting the access
gitpod /workspace/kreypi $ git push
Enumerating objects: 4, done.
Counting objects: 100% (4/4), done.
Delta compression using up to 16 threads
Compressing objects: 100% (2/2), done.
Writing objects: 100% (3/3), 272 bytes | 272.00 KiB/s, done.
Total 3 (delta 0), reused 0 (delta 0)
To https://github.com/kreypi/kreypi.git
   fecdfc0..85cc6fa  master -> master
```

---

it's very flustrating to resolve when i work on a new repository for X amount of time only to realize that i can't push changes due to the rejected publickey where without knowing what is the issue assuming provided scenario I believe that it causes confusion and flustrstion.

**Comment by JesterOrNot** [comment]
@Kreyren is Gitpod installed on the organization in question?

**Comment by Kreyren** [comment]
> @Kreyren is Gitpod installed on the organization in question?

Assuming that by installed you mean gitpod granted access in https://github.com/settings/connections/applications/484069277e293e6d2a2a as explained in additional info then the **issue is present if gitpod is not installed**, but i am able to open gitpod through Firefox extension or by prefixing github repository with gitpod.io to get the editor instance which causes said issue with inability to push changes -> Proposing some check to warn the end-user to perform sufficient setup to avoid confusion with gitpod pushing git references.

**Comment by JesterOrNot** [comment]
I apologize I get it  now This sounds like a great idea

**Comment by Kreyren** [comment]
> I apologize I get it now This sounds like a great idea

np I tried to edit the OP to make the issue more obvious.

**Comment by jankeromnes** [comment]
Thanks a lot @Kreyren for reporting this problem and proposing a solution.

> Expected behavior
> 
> Failure to push changes is not expected should prompt up the end-user if gitpod is executed on affected repositories to approve gitpod in access controls to avoid confusion and misconfiguration.

To clarify, when your push fails, doesn't Gitpod already show you a notification taking you to https://gitpod.io/access-control/ ? (I think this is supposed to be the current behavior)

I guess your suggestion is to specifically detect this case and take users to the org admin page to authorize Gitpod.

However, as a developer member of an org, you don't always have admin rights on the org. This means that usually, the solution will be to track down an org admin and ask them to authorize Gitpod.

**Comment by Kreyren** [comment]
> To clarify, when your push fails, doesn't Gitpod already show you a notification taking you to https://gitpod.io/access-control/ ? (I think this is supposed to be the current behavior)

I didn't notice any notification nor warning

> However, as a developer member of an org, you don't always have admin rights on the org. 

Noted, but I think that some kind of warning should be present so that end-user is aware that gitpod is unable to push references and why so that org admin can make sufficient configuration.

**Comment by jankeromnes** [comment]
Hey @AlexTugarev, would you have some time to look into this issue? 🙂
